### ISSUE TITLE
bitcoin-abc-0.14.4 reports other bitcoin-abc-0.14.4 nodes as misbehaving, bans them | tx was not accepted: non-final (code 64)

### ISSUE THREAD

**Issue by c0deright** [Bitcoin-ABC/bitcoin-abc#29]
### Describe the issue
After upgrading a `proxy` node (connected to the internet with a public IP) and a `client` node (connected only to `proxy`, running txindex=1) from bitcoin-core to bitcoin-abc everything is falling apart.

The `client` node is rejecting TX it requested from `proxy` with error **was not accepted: non-final (code 64)**. The `proxy` node receives this reject but seems not to understand `reject` since it logs **Unknown command "reject"**. There seems to be a protocol mismatch between both bitcoin-abc nodes because right after that the client raises misbehaving level for the `proxy` node and finally when misbehaving level for it's only connection to the bitcoin-abc network (node `proxy`) reaches 100 it drops the connection.

#### If so, please list the steps to reproduce below:
1. run bitcoin-core-0.14.2 on node `proxy` in pruning mode
2. run bitcoin-core-0.14.2 on node `client` in full mode (no pruning) with additional options
 `connect=proxy`
`txindex=1`
3. stop bitcoin-core on both nodes and start bitcoin-abc-0.14.4 on both nodes (same datadir, same bitcoin.conf etc)

### Expected behaviour
`client` connects to `proxy` and resumes block download from `proxy`

### Actual behaviour
`client` connects to `proxy` and then reports `proxy` as misbehaving, drops connection and repeats

### What version of bitcoin-core were you using?
0.14.2 release, self compiled on ubuntu xenial.

### What version of bitcoin-abc are you using?
0.14.4 release, self compiled on ubuntu xenial.

### Machine specs:
- OS: Ubuntu 16.04 LTS
- AWS t2.small instance
- RAM: 2G
- Disk size: 180G
- Disk Type (HD/SDD): SSD

### Any extra information that might be useful in the debugging process.
`debug.log` from `client`:

```
2017-07-26 06:11:36 trying connection <IP-FROM-NODE-PROXY>:8333 lastseen=0.0hrs
2017-07-26 06:11:36 addcon thread start
2017-07-26 06:11:36 net thread start
2017-07-26 06:11:36 Added connection to <IP-FROM-NODE-PROXY>:8333 peer=0
2017-07-26 06:11:36 sending version (113 bytes) peer=0
2017-07-26 06:11:36 send version message: version 70015, blocks=477485, us=[::]:0, them=[::]:0, peer=0
2017-07-26 06:11:37 received: version (113 bytes) peer=0
2017-07-26 06:11:37 sending verack (0 bytes) peer=0
2017-07-26 06:11:37 sending getaddr (0 bytes) peer=0
2017-07-26 06:11:37 receive version message: /Bitcoin ABC:0.14.4(EB8.0)/: version 70015, blocks=477618, us=[::]:0, peer=0, peeraddr=<IP-FROM-NODE-PROXY>:8333
2017-07-26 06:11:37 added time data, samples 2, offset -1 (+0 minutes)
2017-07-26 06:11:37 received: verack (0 bytes) peer=0
2017-07-26 06:11:37 sending sendheaders (0 bytes) peer=0
2017-07-26 06:11:37 sending sendcmpct (9 bytes) peer=0
2017-07-26 06:11:37 sending ping (8 bytes) peer=0
2017-07-26 06:11:37 sending feefilter (8 bytes) peer=0
2017-07-26 06:11:37 received: sendheaders (0 bytes) peer=0
2017-07-26 06:11:37 received: sendcmpct (9 bytes) peer=0
2017-07-26 06:11:37 received: ping (8 bytes) peer=0
2017-07-26 06:11:37 sending pong (8 bytes) peer=0
2017-07-26 06:11:37 received: getheaders (997 bytes) peer=0
2017-07-26 06:11:37 getheaders -1 to end from peer=0
2017-07-26 06:11:37 sending headers (1 bytes) peer=0
2017-07-26 06:11:37 received: pong (8 bytes) peer=0
2017-07-26 06:11:37 received: inv (37 bytes) peer=0
2017-07-26 06:11:37 got inv: tx e6b7abdbc56304b84c1775d32fb594dfc32cec4a3c42cc1faa54ae10dc948866  new peer=0
2017-07-26 06:11:37 askfor tx e6b7abdbc56304b84c1775d32fb594dfc32cec4a3c42cc1faa54ae10dc948866  0 (00:00:00) peer=0
2017-07-26 06:11:37 Requesting tx e6b7abdbc56304b84c1775d32fb594dfc32cec4a3c42cc1faa54ae10dc948866 peer=0
2017-07-26 06:11:37 sending getdata (37 bytes) peer=0
2017-07-26 06:11:37 received: inv (37 bytes) peer=0
2017-07-26 06:11:37 got inv: tx 6eb0e9a491e62b6a1178f605a54387bc7363fe4d90a6110656fb583da70efb1d  new peer=0
2017-07-26 06:11:37 askfor tx 6eb0e9a491e62b6a1178f605a54387bc7363fe4d90a6110656fb583da70efb1d  0 (00:00:00) peer=0
2017-07-26 06:11:37 Requesting tx 6eb0e9a491e62b6a1178f605a54387bc7363fe4d90a6110656fb583da70efb1d peer=0
2017-07-26 06:11:37 sending getdata (37 bytes) peer=0
2017-07-26 06:11:37 received: tx (225 bytes) peer=0
2017-07-26 06:11:37 e6b7abdbc56304b84c1775d32fb594dfc32cec4a3c42cc1faa54ae10dc948866 from peer=0 was not accepted: non-final (code 64)
2017-07-26 06:11:37 sending reject (46 bytes) peer=0
2017-07-26 06:11:37 Misbehaving: <IP-FROM-NODE-PROXY>:8333 peer=0 (0 -> 10)
2017-07-26 06:11:37 received: inv (37 bytes) peer=0
[..]
2017-07-26 06:11:39 got inv: tx fb80f5fda7e8fa66b3f23978cfc83918efc5bf7ccfc06a295ddceee1ed7a41de  new peer=0
2017-07-26 06:11:39 askfor tx fb80f5fda7e8fa66b3f23978cfc83918efc5bf7ccfc06a295ddceee1ed7a41de  0 (00:00:00) peer=0
2017-07-26 06:11:39 Requesting tx fb80f5fda7e8fa66b3f23978cfc83918efc5bf7ccfc06a295ddceee1ed7a41de peer=0
[..]
2017-07-26 06:11:39 received: tx (373 bytes) peer=0
2017-07-26 06:11:39 fb80f5fda7e8fa66b3f23978cfc83918efc5bf7ccfc06a295ddceee1ed7a41de from peer=0 was not accepted: non-final (code 64)
2017-07-26 06:11:39 sending reject (46 bytes) peer=0
2017-07-26 06:11:39 Misbehaving: <IP-FROM-NODE-PROXY>:8333 peer=0 (50 -> 60)
[..]
2017-07-26 06:11:39 received: tx (520 bytes) peer=0
2017-07-26 06:11:39 8ba23b3d23dc69642e7c47ad3f81c4a797c55fa1cf81292e61ae0efb6640b0a0 from peer=0 was not accepted: non-final (code 64)
2017-07-26 06:11:39 sending reject (46 bytes) peer=0
2017-07-26 06:11:39 Misbehaving: <IP-FROM-NODE-PROXY>:8333 peer=0 (80 -> 90)
2017-07-26 06:11:39 received: notfound (37 bytes) peer=0
2017-07-26 06:11:39 received: notfound (37 bytes) peer=0
2017-07-26 06:11:39 received: notfound (73 bytes) peer=0
2017-07-26 06:11:40 received: inv (37 bytes) peer=0
2017-07-26 06:11:40 got inv: tx dc4804fac9a0c8415c5516c5f4855cf6fd65fdfe11c51c54920ad16f780b9e0d  new peer=0
2017-07-26 06:11:40 askfor tx dc4804fac9a0c8415c5516c5f4855cf6fd65fdfe11c51c54920ad16f780b9e0d  0 (00:00:00) peer=0
2017-07-26 06:11:40 Requesting tx dc4804fac9a0c8415c5516c5f4855cf6fd65fdfe11c51c54920ad16f780b9e0d peer=0
2017-07-26 06:11:40 sending getdata (37 bytes) peer=0
2017-07-26 06:11:40 received: inv (37 bytes) peer=0
2017-07-26 06:11:40 got inv: tx 3f15c20667c0f130a9adb3c19d1013cc6c837cbb4239b086ae32dd33d0ad9d15  new peer=0
2017-07-26 06:11:40 askfor tx 3f15c20667c0f130a9adb3c19d1013cc6c837cbb4239b086ae32dd33d0ad9d15  0 (00:00:00) peer=0
2017-07-26 06:11:40 Requesting tx 3f15c20667c0f130a9adb3c19d1013cc6c837cbb4239b086ae32dd33d0ad9d15 peer=0
2017-07-26 06:11:40 sending getdata (37 bytes) peer=0
2017-07-26 06:11:40 received: tx (373 bytes) peer=0
2017-07-26 06:11:40 dc4804fac9a0c8415c5516c5f4855cf6fd65fdfe11c51c54920ad16f780b9e0d from peer=0 was not accepted: non-final (code 64)
2017-07-26 06:11:40 sending reject (46 bytes) peer=0
2017-07-26 06:11:40 Misbehaving: <IP-FROM-NODE-PROXY>:8333 peer=0 (90 -> 100) BAN THRESHOLD EXCEEDED
2017-07-26 06:11:40 disconnecting peer=0
```

`debug.log` from `proxy`:

```
2017-07-26 06:11:36 Added connection to <IP-FROM-NODE-CLIENT>:53654 peer=6
2017-07-26 06:11:36 connection from <IP-FROM-NODE-CLIENT>:53654 accepted
2017-07-26 06:11:36 received: version (113 bytes) peer=6
2017-07-26 06:11:36 sending version (113 bytes) peer=6
2017-07-26 06:11:36 send version message: version 70015, blocks=477618, us=[::]:0, them=[::]:0, peer=6
2017-07-26 06:11:36 sending verack (0 bytes) peer=6
2017-07-26 06:11:36 receive version message: /Bitcoin ABC:0.14.4(EB8.0)/: version 70015, blocks=477485, us=[::]:0, peer=6, peeraddr=<IP-FROM-NODE-CLIENT>:53654
2017-07-26 06:11:36 added time data, samples 8, offset +0 (+0 minutes)
2017-07-26 06:11:37 received: verack (0 bytes) peer=6
2017-07-26 06:11:37 sending sendheaders (0 bytes) peer=6
2017-07-26 06:11:37 sending sendcmpct (9 bytes) peer=6
2017-07-26 06:11:37 sending ping (8 bytes) peer=6
2017-07-26 06:11:37 initial getheaders (477617) to peer=6 (startheight:477485)
2017-07-26 06:11:37 sending getheaders (997 bytes) peer=6
2017-07-26 06:11:37 received: getaddr (0 bytes) peer=6
2017-07-26 06:11:37 received: sendheaders (0 bytes) peer=6
2017-07-26 06:11:37 received: sendcmpct (9 bytes) peer=6
2017-07-26 06:11:37 received: ping (8 bytes) peer=6
2017-07-26 06:11:37 sending pong (8 bytes) peer=6
2017-07-26 06:11:37 received: feefilter (8 bytes) peer=6
2017-07-26 06:11:37 received: feefilter of 0.00001000 BTC/kB from peer=6
2017-07-26 06:11:37 received: pong (8 bytes) peer=6
2017-07-26 06:11:37 received: headers (1 bytes) peer=6
2017-07-26 06:11:37 connection to 46.243.172.188:8333 timeout
2017-07-26 06:11:37 sending inv (37 bytes) peer=4
2017-07-26 06:11:37 received: inv (1261 bytes) peer=4
[..]
2017-07-26 06:11:37 sending tx (224 bytes) peer=6
2017-07-26 06:11:37 sending inv (37 bytes) peer=6
2017-07-26 06:11:37 received: tx (667 bytes) peer=4
2017-07-26 06:11:37 AcceptToMemoryPool: peer=4: accepted 0cb953937659efc1dbbfffad8da5b06ee0241aed3beb54b0ab18651417f5afe2 (poolsz 460 txn, 1317 kB)
2017-07-26 06:11:37 received: reject (46 bytes) peer=6
2017-07-26 06:11:37 Reject tx code 64: non-final: hash e6b7abdbc56304b84c1775d32fb594dfc32cec4a3c42cc1faa54ae10dc948866
2017-07-26 06:11:37 Unknown command "reject" from peer=6
2017-07-26 06:11:37 sending inv (37 bytes) peer=6
2017-07-26 06:11:37 received: getdata (37 bytes) peer=6
2017-07-26 06:11:37 received getdata (1 invsz) peer=6
2017-07-26 06:11:37 received getdata for: tx 1ce0478e96462b26e3ead53650e02f3b4d43582a5fc542df99c9fc38fd2595c1 peer=6
2017-07-26 06:11:37 sending tx (260 bytes) peer=6
2017-07-26 06:11:37 received: reject (46 bytes) peer=6
2017-07-26 06:11:37 Reject tx code 64: non-final: hash 6eb0e9a491e62b6a1178f605a54387bc7363fe4d90a6110656fb583da70efb1d
2017-07-26 06:11:37 Unknown command "reject" from peer=6
[..]
```

**Comment by c0deright** [comment]
There was a strange way to fix this:

Remove `connect=proxy` from `client`'s bitcoin.conf and restart bitcoin-abc so `client` downloads missing blocks from the internet directly (still reporting many misbehaving clients on the internet in debug.log).

After that revert the change in bitcoin.conf so `client` only connects to `proxy` after restart.

Since then no problems, no reports of `proxy` misbehaving in `client`'s debug.log.

This is strange.

**Comment by deadalnix** [comment]
Is this on testnet or main net ?

**Comment by c0deright** [comment]
main net, of course.

This was a production bitcoin core node that got snapshotted (node shutdown, volume snapshots, ...) and from that snapshot a clone was created. then core was replaced with abc 0.14.4 and the problems began.

**Comment by deadalnix** [comment]
Am I correct assuming that client only connect to the network through proxy and do not connect to any other node ?

**Comment by c0deright** [comment]
Right. `client` had `connect=<IP FROM PROXY>` in it's `bitcoin.conf` so it would only connect to `proxy` with only 1 connection. No connection to any other node.

**Comment by deadalnix** [comment]
I just reproduced the set up and will investigate what's going on. Thanks for reporting.

**Comment by c0deright** [comment]
I don't know whether the differnce in block height plays a role in this scenario.

`proxy` was at 477618, `client` was back at 477485.

Thanks for looking into this.

**Comment by deadalnix** [comment]
I was able to repro this once, but cannot anymore. There is indeed something fishy hiding in there. I'll keep digging to get to the bottom of this.

**Comment by deadalnix** [comment]
Ok, I identified the root cause of that problem. I'm working on a patch now.

**Comment by c0deright** [comment]
I also observed another buggy behaviour with abc-0.14.5 yesterday (upgraded all abc nodes):

* `proxy` (abc-0.14.5) was at block 477666, running with `prune=201480` as mentioned above
* `client1` was at block 477485 when I stopped bitcoin-core on that machine and started abc-0.14.5
* `client2` was at block 477485 when I stopped bitcoin-core on that machine and started abc-0.14.5
* `client1` and `client2` are running with `connect=proxy`, no txindex, full node

Both `client1` and `client2` did *not* replicate the missing blocks from `proxy`.

I waited until `proxy` got some new blocks (`getinfo` would return 477667, debug.log had `UpdateTip: new best=[..] height=477667` added) but the clients never started downloading the missing blocks starting at 477486.

Like in the workaround mentioned above I had to disable the `connect` statement in `client1` and `client2` so they would download the blocks from (I assume: non-Bitcoin-ABC nodes) nodes from the internet.

After the clients had the same block height as `proxy` I enabled the `connect=proxy` line again in their bitcoin.conf. Subsequent blocks were downloaded without issues from `proxy` by the clients.

Is this observed behaviour related this issue?
